### PR TITLE
crypto: add sr25519 as a validator key

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

Add sr25519 as a validator key option. We support the crypto in tendermint and added batch verification recently.

### CODE REVIEW

File: crypto/encoding/codec.go [hunk 1]
```diff
@@ -57,6 +64,14 @@ func PubKeyFromProto(k pc.PublicKey) (crypto.PubKey, error) {
 		pk := make(secp256k1.PubKey, secp256k1.PubKeySize)
 		copy(pk, k.Secp256K1)
 		return pk, nil
+	case *pc.PublicKey_Sr25519:
+		if len(k.Sr25519) != secp256k1.PubKeySize {
```
Comment (marbar3778) [inline crypto/encoding/codec.go:68]:
```suggestion
		if len(k.Sr25519) != sr25519.PubKeySize {
```

File: crypto/encoding/codec.go [hunk 2]
```diff
@@ -57,6 +64,14 @@ func PubKeyFromProto(k pc.PublicKey) (crypto.PubKey, error) {
 		pk := make(secp256k1.PubKey, secp256k1.PubKeySize)
 		copy(pk, k.Secp256K1)
 		return pk, nil
+	case *pc.PublicKey_Sr25519:
+		if len(k.Sr25519) != sr25519.PubKeySize {
+			return nil, fmt.Errorf("invalid size for PubKeySecp256k1. Got %d, expected %d",
+				len(k.Sr25519), secp256k1.PubKeySize)
```
Comment (webmaster128) [inline crypto/encoding/codec.go:70]:
```suggestion
				len(k.Sr25519), sr25519.PubKeySize)
```

File: types/params.go [hunk 3]
```diff
@@ -24,11 +25,13 @@ const (
 
 	ABCIPubKeyTypeEd25519   = ed25519.KeyType
 	ABCIPubKeyTypeSecp256k1 = secp256k1.KeyType
+	ABCIPubKeyTypeSr255519  = sr25519.KeyType
 )
 
 var ABCIPubKeyTypesToNames = map[string]string{
 	ABCIPubKeyTypeEd25519:   ed25519.PubKeyName,
 	ABCIPubKeyTypeSecp256k1: secp256k1.PubKeyName,
+	ABCIPubKeyTypeSr255519:  sr25519.PubKeyName,
```
Comment (webmaster128) [inline types/params.go:34]:
```suggestion
	ABCIPubKeyTypeSr25519:  sr25519.PubKeyName,
```

File: types/params.go [hunk 4]
```diff
@@ -24,11 +25,13 @@ const (
 
 	ABCIPubKeyTypeEd25519   = ed25519.KeyType
 	ABCIPubKeyTypeSecp256k1 = secp256k1.KeyType
+	ABCIPubKeyTypeSr255519  = sr25519.KeyType
```
Comment (webmaster128) [inline types/params.go:28]:
```suggestion
	ABCIPubKeyTypeSr25519  = sr25519.KeyType
```

File: types/params_test.go [hunk 5]
```diff
@@ -14,6 +14,7 @@ import (
 var (
 	valEd25519   = []string{ABCIPubKeyTypeEd25519}
 	valSecp256k1 = []string{ABCIPubKeyTypeSecp256k1}
+	valSr25519   = []string{ABCIPubKeyTypeSr255519}
```
Comment (webmaster128) [inline types/params_test.go:17]:
```suggestion
	valSr25519   = []string{ABCIPubKeyTypeSr25519}
```

**Review by webmaster128 — COMMENTED** [review]
Nice, thank you

File: CHANGELOG_PENDING.md [hunk 6]
```diff
@@ -53,6 +53,7 @@ Friendly reminder: We have a [bug bounty program](https://hackerone.com/tendermi
 
 - [config] Add `--mode` flag and config variable. See [ADR-52](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-052-tendermint-mode.md) @dongsam
 - [rpc] /#6329 Don't cap page size in unsafe mode (@gotjoshua, @cmwaters)
+- [crypto] /#6376 Enable sr25519 as a validator key
```
Comment (webmaster128) [inline CHANGELOG_PENDING.md:56]:
```suggestion
- [rpc] \#6329 Don't cap page size in unsafe mode (@gotjoshua, @cmwaters)
- [crypto] \#6376 Enable sr25519 as a validator key
```

File: types/params_test.go [hunk 7]
```diff
@@ -129,7 +130,25 @@ func TestConsensusParamsUpdate(t *testing.T) {
 			},
 			makeParams(100, 200, 300, 50, valSecp256k1),
 		},
+		{
+			makeParams(1, 2, 3, 0, valEd25519),
```
Comment (webmaster128) [inline types/params_test.go:134]: I'm not familiar with this code, but this looks like a copy/paste bug

Comment (marbar3778) [reply types/params_test.go:134]: This is testing the update functionality. It is fine as is.

File: CHANGELOG_PENDING.md (hunk 6 above)
Comment (marbar3778) [reply CHANGELOG_PENDING.md]: thanks

File: abci/types/pubkey.go [hunk 8]
```diff
@@ -38,6 +39,17 @@ func UpdateValidator(pk []byte, power int64, keyType string) ValidatorUpdate {
...
 			Power:  power,
 		}
+	case sr25519.KeyType:
+		pke := sr25519.PubKey(pk)
+		pkp, err := cryptoenc.PubKeyToProto(pke)
+		if err != nil {
+			panic(err)
+		}
+		return ValidatorUpdate{
+			// Address:
```
Comment (alexanderbez) [inline abci/types/pubkey.go:49]: Do we need to set address? If not, let's remove the comment.

File: crypto/encoding/codec.go [hunk 9]
```diff
@@ -57,6 +64,14 @@ func PubKeyFromProto(k pc.PublicKey) (crypto.PubKey, error) {
 		pk := make(secp256k1.PubKey, secp256k1.PubKeySize)
 		copy(pk, k.Secp256K1)
 		return pk, nil
+	case *pc.PublicKey_Sr25519:
+		if len(k.Sr25519) != sr25519.PubKeySize {
+			return nil, fmt.Errorf("invalid size for PubKeySecp256k1. Got %d, expected %d",
```
Comment (alexanderbez) [inline crypto/encoding/codec.go:69]:
```suggestion
			return nil, fmt.Errorf("invalid size for PubKeySR25519; got %d, expected %d",
```

File: types/params.go [hunk 10]
```diff
@@ -24,11 +25,13 @@ const (
 
 	ABCIPubKeyTypeEd25519   = ed25519.KeyType
 	ABCIPubKeyTypeSecp256k1 = secp256k1.KeyType
+	ABCIPubKeyTypeSr25519  = sr25519.KeyType
```
Comment (alexanderbez) [inline types/params.go:28]: format document :)

**Review by alexanderbez — CHANGES_REQUESTED** [review]

**Review by alexanderbez — APPROVED** [review]

**Comment by cmwaters** [comment]
Does this test failing have anything to do with your changes:
```
--- FAIL: TestNonEd25519Pubkey (0.01s)
    secret_connection_test.go:282: 
        	Error Trace:	secret_connection_test.go:282
        	Error:      	An error is expected but got nil.
        	Test:       	TestNonEd25519Pubkey
FAIL
```

**Pull request merged**